RoFILib/configuration/Configuration.cpp

```diff
@@ -0,0 +1,496 @@
+
+        ID idNext = edge.id2();
+        ShoeId sideNext = edge.side2();
+
+        // Compute, where next module should be.
+        Matrix matrixCmp = computeConnectedMatrix(edge);
+
+        if (matrices.find(idNext) == matrices.end()) {
+            matrices[idNext][sideNext] = matrixCmp;
+            result &= computeMatricesRec(idNext, sideNext);
+        } else {
+            const auto& matrixNext = matrices.at(idNext)[sideNext];
+            if (!equals(matrixNext, matrixCmp))
+                return false;
+        }
+    }
+    return result;
+}
+
+Matrix Configuration::computeOtherSideMatrix(ID id, ShoeId side) const {
+    auto const& mod = modules.at(id);
+    auto const& matrix = matrices.at(id)[side];
+    double alpha = 2 * M_PI * mod.alpha / 360;
+    double beta = 2 * M_PI * mod.beta / 360;
+    double gamma = 2 * M_PI * mod.gamma / 360;
+
+    if (side == B)
+        std::swap(alpha, beta);
+
+    return matrix * transformJoint(alpha, beta, gamma);
+}
+
+Matrix Configuration::computeConnectedMatrix(Edge edge) const {
+    auto const& matrix = matrices.at(edge.id1())[edge.side1()];
+    return matrix * transformConnection(edge.dock1(), edge.ori(), edge.dock2());
+}
+
+bool Configuration::connected() const {
+    std::unordered_set<ID> seen;
+    seen.insert(fixedId);
+    dfsID(fixedId, seen);
+    return seen.size() == modules.size();
+}
+
+void Configuration::dfsID(ID id, std::unordered_set<ID>& seen) const {
+    for (const std::optional<Edge> edgeOpt : edges.at(id)) {
+        if (!edgeOpt.has_value())
+            continue;
+        ID nextId = edgeOpt.value().id2();
+        if (seen.find(nextId) == seen.end()) {
+            seen.insert(nextId);
+            dfsID(nextId, seen);
+        }
+    }
+}
+
+bool Configuration::collisionFree() const {
+    for (auto it1 = matrices.begin(); it1 != matrices.end(); ++it1) {
+        for (auto it2 = it1; it2 != matrices.end(); ++it2) {
+            const auto& ms1 = it1->second;
+            const auto& ms2 = it2->second;
+            if (it1 == it2) {
+                if (centerSqDistance(ms1[A], ms1[B]) < 1)
+                    return false;
+                continue;
+            } 
+            if (centerSqDistance(ms1[A], ms2[A]) < 1 ||
+                centerSqDistance(ms1[B], ms2[B]) < 1 ||
+                centerSqDistance(ms1[B], ms2[A]) < 1 ||
+                centerSqDistance(ms1[A], ms2[B]) < 1)
+                return false;
+        }
+    }
+    return true;
+}
+
+Vector Configuration::massCenter() const {
+    Vector mass({0,0,0,1});
+    for (const auto& [id, ms] : matrices) {
+        mass += center(ms[A]);
+        mass += center(ms[B]);
+    }
+    mass /= modules.size()*2;
+    return mass;
+}
+
+bool Configuration::execute(const Action& action) {
+    bool ok = true;
+    for (const Action::Rotate rot : action.rotations())
+        ok &= execute(rot);
+
+    for (const Action::Reconnect rec : action.reconnections())
+        ok &= execute(rec);
```

Comment:
I am not sure, how it is used in Rofilib/distribute and in rrt, but in other parts of repo it would make sense.
So should I change it?

Reply:
I would change it - but consult it with tests, please.

Reply:
Seems that `distribute` doesn't have tests.

Reply:
Let's keep it as it this then. I don't want to break `distribute` as there is no one to fix it and bug finding in the future might be hard.

Reply:
It might be worth testing the impact of this change on efficiency of the algorithms based on astar, but I'll leave it as it this for now. Could you open issue about this?